### PR TITLE
ECO-1047 JSON Serialization

### PR DESCRIPTION
**Pull request by hoffmannjan**

### CODE REVIEW

**Comment by hoffmannjan** [comment]
Includes and closes #21

File: src/lib/CLValue2/Abstract.ts [hunk 1]
```diff
@@ -1,13 +1,70 @@
...
+  }
 }
 
-export abstract class ToBytes {
-  toBytes: () => Uint8Array;
+export abstract class ToBytes {}
+
+export abstract class FromBytes {}
+
+export interface ResultAndRemainder<T, E> {
```
Comment (zie1ony) [inline src/lib/CLValue2/Abstract.ts:71]: That could be `FromBytesResult`.

File: src/lib/CLValue2/AccountHash.ts [hunk 2]
```diff
@@ -1,13 +1,30 @@
-import { CLType, CLValue, ToBytes } from './Abstract';
+// NOTE: Currently this isn't supported CLValue
```
Comment (zie1ony) [inline src/lib/CLValue2/AccountHash.ts:1]: Remove `AccountHash`. We will reintroduce that later.

File: src/lib/CLValue2/List.test.ts [hunk 3]
```diff
@@ -120,8 +135,57 @@ describe('CLValue List implementation', () => {
...
     const myList = new CLList([new CLBool(false)]);
-    const expected = Uint8Array.from([ 1, 0, 0, 0, 0 ]);
-    expect(myList.toBytes()).to.deep.eq(expected);
+    const expected = Uint8Array.from([1, 0, 0, 0, 0]);
+    expect(myList.toBytes().unwrap()).to.deep.eq(expected);
   });
-});
 
+  it('Runs fromBytes properly', () => {
+    const myList = new CLList([new CLBool(false), new CLBool(true)]);
```
Comment (zie1ony) [inline src/lib/CLValue2/List.test.ts:140]: No need to have every line separated with new line.

File: src/lib/CLValue2/constants.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+export enum KeyVariant {
+  Account,
+  Hash,
+  URef
+}
+
+export const BOOL_CL_TYPE = 'Bool';
+
+export type CLTypes = typeof BOOL_CL_TYPE;
```
Comment (zie1ony) [inline src/lib/CLValue2/constants.ts]: What's that?

File: src/lib/CLValue2/utils.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  CLU64Type,
+  CLU128Type,
+  CLU256Type,
+  CLU512Type,
+  CLTuple1Type,
+  CLTuple2Type,
+  CLTuple3Type
+} from './index';
+
+// const cl_type = { List: { List: 'Bool' } };
```
Comment (zie1ony) [inline src/lib/CLValue2/utils.ts]: Remove that comment.

File: test/lib/RuntimeArgs.test.ts [hunk 6]
```diff
@@ -66,6 +66,21 @@ describe(`RuntimeArgs`, () => {
     );
   });
 
+//   it('should deserialize Option of U512', () => {
```
Comment (zie1ony) [inline test/lib/RuntimeArgs.test.ts]: Why is that commented?

File: src/lib/CLValue2/Abstract.ts [hunk 7]
```diff
@@ -1,13 +1,70 @@
...
+}
+
+export const resultHelper = <T, E>(
+  arg1: Result<T, E>,
+  arg2?: Uint8Array
+): ResultAndRemainder<T, E> => {
+  return { result: arg1, remainder: arg2 };
+};
+
+export interface CLJsonFormat {
```
Comment (zie1ony) [inline src/lib/CLValue2/Abstract.ts]: That should be `CLValueJsonFormat`.

**Review by zie1ony — APPROVED** [review]
Looks good.

**Pull request merged**